File: rust/src/bfs.rs
use super::array2d::Array2D;
use super::{Point, PointCoord};
use std::cell::RefCell;

const OFFSETS: &[(isize, isize)] = &[(1, 0), (-1, 0), (0, 1), (0, -1)];

fn neighbors(pos: Point) -> impl Iterator<Item = Point> {
    #[cfg(not(feature = "ignore-negative-coords"))]
    return OFFSETS.iter().filter_map(move |offset| {
        let x = pos.0 as i32 + offset.0 as i32;
        let y = pos.1 as i32 + offset.1 as i32;
        if x < 0 || y < 0 {
            return None;
        }
        Some(Point(x as PointCoord, y as PointCoord))
    });
    #[cfg(feature = "ignore-negative-coords")]
    return OFFSETS.iter().map(move |offset| {
        Point(
            (pos.0 as i32 + offset.0 as i32) as PointCoord,
            (pos.1 as i32 + offset.1 as i32) as PointCoord,
        )
    });
}

struct State {
    visited: Array2D<bool>,
    depth: Array2D<i16>,
}

impl State {
    fn new(width: usize, height: usize) -> Self {
        Self {
            visited: Array2D::filled_with(false, width, height),
            depth: Array2D::filled_with(-1, width, height),
        }
    }
    fn clear(&mut self) {
        self.visited.fill(false);
        self.depth.fill(-1);
    }
}

pub struct BFS {
    width: usize,
    height: usize,
    walls: Array2D<bool>,
    #[cfg(feature = "alloc-state-once")]
    state: RefCell<State>,
}

impl BFS {
    pub fn new(width: usize, height: usize) -> Self {
        Self {
            width,
            height,
            walls: Self::generate_walls(width, height),
            #[cfg(feature = "alloc-state-once")]
            state: RefCell::new(State::new(width, height)),
        }
    }

    fn generate_walls(width: usize, height: usize) -> Array2D<bool> {
        let mut walls = Array2D::filled_with(false, width, height);

        for index in 0..width as PointCoord {
            walls[Point(index, 0)] = true;
            walls[Point(index, height as PointCoord - 1)] = true;
        }

        for index in 0..height as PointCoord {
            walls[Point(0, index)] = true;
            walls[Point(width as PointCoord - 1, index)] = true;
        }

        let h = height / 10;
        let w = width / 10;

        for index in 0..height - h {
            let x = 2 * w;
            let y = index;
            walls[Point(x as PointCoord, y as PointCoord)] = true;
        }

        for index in h..height {
            let x = 8 * w;
            let y = index;
            walls[Point(x as PointCoord, y as PointCoord)] = true;
        }

        walls
    }

    pub fn path(&self, from: Point, to: Point) -> Option<Vec<Point>> {
        #[cfg(feature = "alloc-state-once")]
        let mut state = {
            let mut state = self.state.borrow_mut();
            state.clear();
            state
        };
        #[cfg(not(feature = "alloc-state-once"))]
        let mut state = State::new(self.width, self.height);

        state.visited[from] = true;
        state.depth[from] = 0;

        let mut queue = std::collections::VecDeque::with_capacity(self.width * self.height);
        queue.push_back(from);
        while let Some(pos) = queue.pop_front() {
            let length = state.depth[pos];
            if pos == to {
                break;
            }

            for new_pos in neighbors(pos) {
                if state.visited[new_pos] || self.walls[new_pos] {
                    continue;
                }
                state.visited[new_pos] = true;
                state.depth[new_pos] = length + 1;
                queue.push_back(new_pos);
            }
        }

        // not found
        if !state.visited[to] {
            return None;
        }

        let mut pos = to;
        let mut result = Vec::with_capacity(state.depth[pos] as usize);
        result.push(pos);
        while pos != from {
            let length = state.depth[pos];
            for prev_pos in neighbors(pos) {
                if state.depth[prev_pos] == length - 1 {
                    pos = prev_pos;
                    result.push(pos);
                    break; // push first found point
                }
            }
        }

        result.reverse();
        Some(result)
    }
}

File: kotlin/src/BFS.kt
@file:Suppress("NOTHING_TO_INLINE")

import java.util.*
import kotlin.collections.ArrayDeque
import kotlin.collections.ArrayList

class Point(val x: Int, val y: Int)

private const val DEFAULT_DEPTH: Short = -1
private const val DEFAULT_QUEUE: Short = 0

class BFS(val width: Int, val height: Int) {
    private val walls = BooleanArray(width * height) { false }

    val cachedPoints = Array(width * height) { Point(it % width, it / width) }

    val depth = ShortArray(width * height) { DEFAULT_DEPTH }
    val queue = ShortArray(width * height) { DEFAULT_QUEUE }

    fun generateWalls() {
        for (index in 0 until width * height) {
            walls[index] = false
        }
        for (index in 0 until width) {
            walls[index] = true
            walls[index + (height - 1) * width] = true
        }
        for (index in 0 until height) {
            walls[index * width] = true
            walls[width - 1 + index * width] = true
        }

        val h = height / 10
        val w = width / 10
        for (index in 0 until height - h) {
            val x = 2 * w
            val y = index
            walls[x + y * width] = true
        }
        for (index in h until height) {
            val x = 8 * w
            val y = index
            walls[x + y * width] = true
        }
    }


    fun path(from: Point, to: Point): Array<Point>? {
        // for optimize use index not Point.
        val fromIndex = getIndex(from.x, from.y)
        val toIndex = getIndex(to.x, to.y)
        val offsets = shortArrayOf(1, -1, width.toShort(), (-width).toShort())  //performs better as local variable

        // fill use bfs
        depth.setAll(DEFAULT_DEPTH)
        depth[fromIndex] = 0

        queue.setAll(DEFAULT_QUEUE)
        queue[0] = fromIndex.toShort()
        var queueIter = 0
        var queueEnd = 1

        while (queueIter < queueEnd) {
            val index = queue[queueIter].toInt()
            if (index == toIndex) {
                break
            }
            queueIter += 1

            val nLength = (depth[index] + 1).toShort()
            for (offset in offsets) {
                val nIndex = index + offset
                if (depth[nIndex] >= 0 || walls[nIndex]) {
                    continue
                }
                depth[nIndex] = nLength
                queue[queueEnd] = nIndex.toShort()
                queueEnd += 1
            }
        }

        if (queueIter == queueEnd) { // not found
            return null
        }

        // make path

        val pathLength = depth[toIndex] + 1
        val result = Array(pathLength) { to }
        var resultIndex = pathLength - 2

        var index = toIndex
        while (index != fromIndex) {
            val nLength = (depth[index] - 1).toShort()

            for (offset in offsets) {
                val nIndex = index + offset
                if (depth[nIndex] == nLength) {
                    index = nIndex
                    result[resultIndex--] = cachedPoints[index]
                    break // push first found point.
                }
            }
        }

        return result
    }

    var openNodes = ArrayDeque<Point>()
    var cachedScore = ShortArray(width * height, { Short.MAX_VALUE })

    fun path2(from: Point, to: Point): ArrayList<Point>? {

        openNodes.clear()
        cachedScore.setAll(Short.MAX_VALUE)
        openNodes.add(from)
        cachedScore.set(getIndex(from), 0)

        while (openNodes.isNotEmpty()) {

            val currentNode = openNodes.removeFirst()
            if (currentNode == to) {
                break
            }
            val currentScore: Short = cachedScore[getIndex(currentNode.x, currentNode.y)]
            forEachSide(currentNode) { x, y ->
                val newScore: Short = (currentScore + 1).toShort()
                if (newScore < cachedScore[getIndex(x, y)]) {
                    cachedScore[getIndex(x, y)] = newScore
                    openNodes.add(getPoint(x, y))
                }
            }
        }

        if (cachedScore.get(getIndex(to)) == Short.MAX_VALUE) { // not found
            return null
        }

        // make path

        val result = ArrayList<Point>()
        var curPoint = to
        result.add(to)

        while (curPoint != from) {
            var minScore = Short.MAX_VALUE
            var nextPoint = curPoint

            forEachSide(curPoint) { x, y ->
                val newScore = cachedScore[getIndex(x, y)]
                if (minScore > newScore) {
                    minScore = newScore
                    nextPoint = cachedPoints[getIndex(x, y)]
                }
            }
            result.add(nextPoint)

            curPoint = result.last()
        }

        result.reverse()
        return result
    }

    private fun getIndex(point: Point): Int {
        return getIndex(point.x, point.y)
    }

    private inline fun forEachSide(currentNode: Point, operation: (x: Int, y: Int) -> Unit) {
        val x = currentNode.x
        val y = currentNode.y
        ifFreeCall(x + -1, y + 0, operation)
        ifFreeCall(x + 1, y + 0, operation)
        ifFreeCall(x + 0, y + 1, operation)
        ifFreeCall(x + 0, y + -1, operation)
    }

    private inline fun ifFreeCall(x: Int, y: Int, operation: (x: Int, y: Int) -> Unit) {
        if (x >= width || x < 0) {
            return
        }
        if (y >= width || y < 0) {
            return
        }

        if (walls[getIndex(x, y)]) {
            return
        }

        operation(x, y)
    }


    inline fun getPoint(x: Int, y: Int): Point {
        return cachedPoints[getIndex(x, y)]
    }

    inline fun getIndex(x: Int, y: Int) = x + y * width
}

private inline fun ShortArray.setAll(value: Short) {
    repeat(size) {
        set(it, value)
    }
}
